module/dummy_psu.c

```diff
@@ -9,9 +9,390 @@
+		return POWER_SUPPLY_USB_TYPE_DCP;
+	};
+	if (!strcmp("POWER_SUPPLY_USB_TYPE_CDP", str_enum)) {
+		return POWER_SUPPLY_USB_TYPE_CDP;
+	};
+	if (!strcmp("POWER_SUPPLY_USB_TYPE_ACA", str_enum)) {
+		return POWER_SUPPLY_USB_TYPE_ACA;
+	};
+	if (!strcmp("POWER_SUPPLY_USB_TYPE_C", str_enum)) {
+		return POWER_SUPPLY_USB_TYPE_C;
+	};
+	if (!strcmp("POWER_SUPPLY_USB_TYPE_PD", str_enum)) {
+		return POWER_SUPPLY_USB_TYPE_PD;
+	};
+	if (!strcmp("POWER_SUPPLY_USB_TYPE_PD_DRP", str_enum)) {
+		return POWER_SUPPLY_USB_TYPE_PD_DRP;
+	};
+	if (!strcmp("POWER_SUPPLY_USB_TYPE_PD_PPS", str_enum)) {
+		return POWER_SUPPLY_USB_TYPE_PD_PPS;
+	};
+	if (!strcmp("POWER_SUPPLY_USB_TYPE_APPLE_BRICK_ID", str_enum)) {
+		return POWER_SUPPLY_USB_TYPE_APPLE_BRICK_ID;
+	};
+
+	return -1;
+}
+
+static inline _Bool power_supply_is_str_property(enum power_supply_property psp)
+{
+	switch (psp) {
+	case POWER_SUPPLY_PROP_MODEL_NAME:
+	case POWER_SUPPLY_PROP_MANUFACTURER:
+	case POWER_SUPPLY_PROP_SERIAL_NUMBER:
+		return 1;
+	default:
+		break;
+	}
+
+	return 0;
+}
+
+// battery_property_map related picked and adapted from test_power.c
+// power_supply_sysfs.c holds all the values
+struct battery_property_map {
+	int value;
+	char const *key;
+};
+
+static struct battery_property_map map_status[] = {
+	{ POWER_SUPPLY_STATUS_UNKNOWN, "Unknown" },
+	{ POWER_SUPPLY_STATUS_CHARGING, "charging" },
+	{ POWER_SUPPLY_STATUS_DISCHARGING, "discharging" },
+	{ POWER_SUPPLY_STATUS_NOT_CHARGING, "not-charging" },
+	{ POWER_SUPPLY_STATUS_FULL, "full" },
+	{ -1, NULL },
+};
+
+static struct battery_property_map map_health[] = {
+	{ POWER_SUPPLY_HEALTH_UNKNOWN, "Unknown" },
+	{ POWER_SUPPLY_HEALTH_GOOD, "good" },
+	{ POWER_SUPPLY_HEALTH_OVERHEAT, "overheat" },
+	{ POWER_SUPPLY_HEALTH_DEAD, "dead" },
+	{ POWER_SUPPLY_HEALTH_OVERVOLTAGE, "overvoltage" },
+	{ POWER_SUPPLY_HEALTH_UNSPEC_FAILURE, "failure" },
+	{ -1, NULL },
+};
+
+static struct battery_property_map map_technology[] = {
+	{ POWER_SUPPLY_TECHNOLOGY_UNKNOWN, "Unknown" },
+	{ POWER_SUPPLY_TECHNOLOGY_NiMH, "NiMH" },
+	{ POWER_SUPPLY_TECHNOLOGY_LION, "Li-ion" },
+	{ POWER_SUPPLY_TECHNOLOGY_LIPO, "Li-poly" },
+	{ POWER_SUPPLY_TECHNOLOGY_LiFe, "LiFe" },
+	{ POWER_SUPPLY_TECHNOLOGY_NiCd, "NiCd" },
+	{ POWER_SUPPLY_TECHNOLOGY_LiMn, "LiMn" },
+	{ -1, NULL },
+};
+
+static struct battery_property_map map_capacity[] = {
+	{ POWER_SUPPLY_CAPACITY_LEVEL_UNKNOWN, "Unknown" },
+	{ POWER_SUPPLY_CAPACITY_LEVEL_CRITICAL, "Critical" },
+	{ POWER_SUPPLY_CAPACITY_LEVEL_LOW, "Low" },
+	{ POWER_SUPPLY_CAPACITY_LEVEL_NORMAL, "Normal" },
+	{ POWER_SUPPLY_CAPACITY_LEVEL_HIGH, "High" },
+	{ POWER_SUPPLY_CAPACITY_LEVEL_FULL, "Full" },
+	{ -1, NULL },
+};
+
+static int map_get_value(struct battery_property_map *map, const char *key)
+{
+	char buf[MAX_KEYLENGTH];
+	int cr;
+
+	strncpy(buf, key, MAX_KEYLENGTH);
+	buf[MAX_KEYLENGTH - 1] = '\0';
+
+	cr = strnlen(buf, MAX_KEYLENGTH) - 1;
+	if (cr < 0)
+		return -1;
+	if (buf[cr] == '\n')
+		buf[cr] = '\0';
```

Comment:
In what case you get `\n` appended to the property name?

Reply:
I've cherry-picked it from `test_power.c` (specified few lines above in comment) but this case is not verified but possible.